fgpyo/util/inspect.py

```diff
@@ -20,10 +16,86 @@
 else:
     from typing_extensions import TypeAlias
 
-import attr
+import dataclasses
+import functools
+from dataclasses import MISSING as DATACLASSES_MISSING
+from dataclasses import fields as get_dataclasses_fields
+from dataclasses import is_dataclass as is_dataclasses_class
+from enum import Enum
+from functools import partial
+from pathlib import PurePath
+from typing import TYPE_CHECKING
+from typing import Callable
+from typing import Optional
+from typing import TypeVar
 
 import fgpyo.util.types as types
 
+try:
+    import attr
+
+    _use_attr = True
+    from attr import NOTHING as ATTR_NOTHING
+    from attr import fields as get_attr_fields
+    from attr import fields_dict as get_attr_fields_dict
+
+    Attribute = attr.Attribute
+    # dataclasses and attr have internal tokens for missing values, join into a set so that we can
+    # check if a value is missing without knowing the type of backing class
+    MISSING = {DATACLASSES_MISSING, ATTR_NOTHING}
+except ImportError:
+    _use_attr = False
+    attr = None
+    ATTR_NOTHING = None
+    Attribute = TypeVar("Attribute", bound=object)  # type: ignore[misc, assignment]
+
+    # define empty placeholders for getting attr fields as a tuple or dict. They will never be
+    # called because the import failed; but they're here to ensure that the function is defined in
+    # sections of code that don't know if the import was successful or not.
+
+    def get_attr_fields(cls: type) -> Tuple[dataclasses.Field, ...]:  # type: ignore[misc]
+        """Get tuple of fields for attr class. attrs isn't imported so return empty tuple."""
+        return ()
+
+    def get_attr_fields_dict(cls: type) -> Dict[str, dataclasses.Field]:  # type: ignore[misc]
+        """Get dict of name->field for attr class. attrs isn't imported so return empty dict."""
+        return {}
+
+    # for consistency with successful import of attr, create a set for missing values
+    MISSING = {DATACLASSES_MISSING}
+
+if TYPE_CHECKING:
+    from _typeshed import DataclassInstance as DataclassesProtocol
+else:
+
+    class DataclassesProtocol(Protocol):
+        __dataclasses_fields__: Dict[str, dataclasses.Field]
+
+
+if TYPE_CHECKING and _use_attr:
+    from attr import AttrsInstance
+else:
+
+    class AttrsInstance(Protocol):  # type: ignore[no-redef]
+        __attrs_attrs__: Dict[str, Any]
+
+
+def is_attr_class(cls: type) -> bool:  # type: ignore[arg-type]
+    """Return True if the class is an attr class, and False otherwise"""
+    return hasattr(cls, "__attrs_attrs__")
+
+
+MISSING_OR_NONE = {*MISSING, None}
+DataclassesOrAttrClass = Union[DataclassesProtocol, AttrsInstance]
+FieldType: TypeAlias = Union[dataclasses.Field, Attribute]
```

Comment:
I would still document these. It makes sense what they are, but a reader will wonder what they are meant to be used for. Document like this so Sphinx picks it up:

```python
AttrFromType = TypeVar("AttrFromType")
"""TypeVar to allow attr_from to be used with either an attr class or a dataclasses class."""
```